Initial Design document  finalize (#1)
sensein/brainkb-design-document: README.md

The main objective of BrainyPedia is to represent Neuroscience knowledge as a knowledge graph such that it can be used for different downstream tasks, such as making predictions and new inferences in addition to querying and viewing information.  The expected outcome of the BrainyPedia includes the following:

- (Semi-)Automated extraction of Neuroscience knowledge from structured, semi-structured, and unstructured sources, representing the extracted knowledge via KGs.
 
- Visualization of the KGs.

- Provides features to perform different analytics operations over the BrainyPedia KGs.
- (Semi-)automatically validates (e.g., quality) the extracted KGs to ensure that the BrainyPedia KGs are high quality. 
- Provides the ability to ingest data in batch or streaming mode for the automated extraction of KGs.

## Why BrainyPedia?

> - **Limited Availability of Platforms for Integrating Neuroscience Data into Knowledge Graphs:** In fields such as biomedicine, many platforms, e.g., [SPOKE](https://doi.org/10.1093/bioinformatics/btad080) and [CIViC](https://civicdb.org/welcome), exist for  (the construction and maintenance of) large-scale KGs to be accessible. <span style="color: red;">However, such resources are comparatively limited in the domain of neuroscience.</span> [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008), a web-based platform that integrates anatomical, functional, and genetic knowledge, is among the limited number of such resources. [BrainKnow](http://www.brain-knowledge-engine.org/), the most recent platform, is another platform that is designed to [synthesizes and integrates neuroscience knowledge from scientific literature](https://arxiv.org/pdf/2403.04346.pdf). Additionally, projects like [DANDI](https://dandiarchive.org/) are making strides by enabling the publication and sharing of neurophysiology data, but not on KGs. 

Reviewer: something wrong with the first sentence, and ()

Reviewer: perhaps: "Additionally, projects like [DANDI](https://dandiarchive.org/) are making strides by enabling sharing of neurophysiology data together with its metadata"

Author: Thanks!

> - **Lack of Support for Heterogeneous Data Sources:** The current platforms in neuroscience are limited in their ability to handle a diverse range of data sources. For instance, [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008) can only integrate knowledge from 41 databases, whereas [BrainKnow](http://www.brain-knowledge-engine.org/) solely focuses on scientific literature. <span style="color: red;">However, knowledge is not restricted to just databases or scientific literature, and there is a need for platforms that can accommodate a wider variety of sources (e.g., structured, semi-structured and unstructured sources).</span>

Reviewer: perhaps you can say something more about "structured, semi-structured and unstructured sources" and provide the examples?

Author: I've added an example and also a new ingest architecture. Perhaps, it should make things clear, at least in my view. let me know if you feel otherwise.

Thanks

## Principles

### Data Ingestion

> BrainyPedia will support the data from various sources in different formats  (e.g., texts, JSON (JavaScript Object Notation)) for knowledge extraction via the BrainyPedia user interface (UI) and the API endpoints. Both batch and streaming data ingestion modes will be supported.

Reviewer: example of sources?

> ### Schema Flexibility

Reviewer: I would think that you should write something more in this section about the schema and how it will evolve (who will be deciding)

KGs evolve over time. Therefore, BrainyPedia will support this evolution by allowing the addition (or removal) of entities and relationships (or new knowledge).

### Maintainability

> BrainyPedia shall be maintainable, allowing operations such as KG enrichment and validation to be performed easily.

Reviewer: either remove this section, or say something more?

### Curation

> BrainyPedia will allow the community-driven curation of the KGs as well as (semi-) automated extraction and construction of KGs from external sources, e.g, scientific literatures.

Reviewer: I understand that first there is data extraction from the sources and then the curation happens, am I right? I wouldn't mix this and perhaps write something more about the curation (e.g. will curation happen always between sharing new data? who will be curating?)

Reviewer: can you say something more what you mean by (semi-) "

### Accuracy, Completeness and Consistency (ACC)

> BrainyPedia shall ensure the accuracy of the knowledge for which multi-step (semi-) automated validations will be performed. Additionally, checks will also be performed to ensure that the KG triples are complete, i.e., the mandatory information is present. Further to accuracy and completeness, BrainyPedia shall ensure that adding the new facts (or KG triples) will not lead to inconsistency (see figure below) with existing knowledge due to factual errors, data inconsistencies, and incompleteness.

Reviewer: do you know already what will be the "multi-steps"? if yes, write something more

Reviewer: Do you know what will be the mechanism when the "inconsistency" is detected? Will you remove the entire node temporarily and ask for human curation? will you keep the triple that come from "more trusted" source?

Author: I will update the description as we now have devised steps to perform the validation

Author: added description

![](acc.png)

> _Figure 1: KGs. The image on the left shows the original knowledge graph, while the image on the right demonstrates the updated knowledge graph. The green highlighted box indicates new knowledge that has been added, while the red highlighted box indicates any inconsistencies caused by factual changes._

Reviewer: i needed to look at the figure for longer time to understand what error or inconsistency you have in mind. Be nice, and mention that it's about adding a new triple that has a different year in DOB :)
Perhaps you can even plot the entire new graph that you're adding so it's clear that there are two triples that contain the same information and they do not agree with each other.

Author: I have added description in caption following suggestion from Yarik.

Caption reads as follows: "Figure 1: KGs. The image on the left shows the original knowledge graph, while the image on the right demonstrates the updated knowledge graph. The green highlighted box indicates new knowledge that has been added, while the red highlighted box indicates any inconsistencies caused by factual changes, i.e., **incorrect date of birth**."

The ACC process will ensure human-centricity is maintained alongside automated validation.

### Provenance

> To enable trust, the provenance, such as the source of the information and the curators (in the case of manual) of all the information, shall be maintained. The provenance conflict resolution mechanism will also be implemented to ensure the accuracy of the provenance information.

Reviewer: how will you ensure that the source stays available?

Reviewer: Do you mean "The provenance OF conflict resolution"? i.e. when the conflict between triples is detected?

### Querying and Reasoning

> BrainyPedia shall support the KGs' querying and reasoning. It shall also support other downstream analytics tasks, such as link predictions using machine learning techniques.

Reviewer: can you provide example of tasks? can you say more about "ink predictions using machine learning techniques"?

Author: I have updated the description with a figure

### Integration and Interoperability

> To ensure interoperability and ease of integration, BrainyPedia will focus on using standardized ontologies or schemas. However, not all standardized ontologies or schemas are available. In such cases, other schemas or ontologies must be used. To ensure the interoperability, the alignment will be performed where necessary.

Reviewer: ```suggestion
To ensure interoperability and ease of integration, BrainyPedia will focus on using standardized ontologies or schemas whenever possible. The alignment between different ontologies will be performed where necessary.
```

### Minimize Cognitive Burden and Data Fatigue

> As BrainyPedia will also provide features to perform the analytics operation in addition to querying the information (or knowledge), a special emphasis shall be placed on ensuring that the information presented to the user does not cause a cognitive burden and data fatigue.  For example, the figure below (left) places more cognitive burden than on the right.

Reviewer: don't understand the point, but perhaps once you add the figure I will have an idea what you mean

Author: updated

## Other considerations

__Assumption:__ We operate on open-world assumptions (OWA), not closed-world assumptions (CSA). In OWA, we do not make any assumptions about the absence of statements, while in CSA absence of statements would be evaluated as false, i.e., assumed to be false.

## Use cases

- **Knowledge Extraction and Integration:** BrainyPedia extracts knowledge from diverse sources, such as projects and lab meetings, and formats (scientific literature, databases, text, and JSON) into unified KG representation, offering a comprehensive and integrated view of neuroscience knowledge. For example, projects like [DANDI:](https://www.dandiarchive.org/), [BICAN:](https://www.portal.brain-bican.org/), [NeuroLex](https://scicrunch.org/scicrunch/interlex/dashboard) and [ReproNim:](https://www.repronim.org/) store vast amounts of neuroscience knowledge; integrating all the knowledge into a single platform will not only provide an integrated view but also enable new knowledge discovery.

- **Visualization & Analytics:**
- 

## Schema

## Usage Scenario
**Actor:** A

> **Role:** Neuroscientists/Researcher

Reviewer: Perhaps you can simplify it by saying "Neuroscientist A" or ("Neuroscientist Alice/Bob etc.)

Author: updated

- **Lack of Support for Heterogeneous Data Sources:** The current platforms in neuroscience are limited in their ability to handle a diverse range of data sources. For instance, [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008) can only integrate knowledge from 41 databases, whereas [BrainKnow](http://www.brain-knowledge-engine.org/) solely focuses on scientific literature. <span style="color: red;">However, knowledge is not restricted to just databases or scientific literature, and there is a need for platforms that can accommodate a wider variety of sources (e.g., structured, semi-structured and unstructured sources).</span>

	- **Example:**
		- Structured data: Relational database
		- Semi-structured data: JSON, XML and CSV files
		- Unsuctured data: Text, Video files

> ## Principles

Reviewer 2: are all of the below principles, or feature sets? i would separate those two considerations. principles should be guiding, and perhaps as precise as could be implemented.

Reviewer: I agree that the division between Principle, Features  and Use Cases is not clear to me. Right now  the Feature section is empty if I understand the format correctly, but some points from Principles and Use Case could be features?

Reviewer: For example "Curation" from "Principles" and "Human in the loop" in "Use cases" are more or less about the same feature, right?

Author: @djarecka @satra I have updated the principles from the KB slide. Previous one could go as a features and have moved to features section.

- **Doing meta-analysis:** Meta-analysis is a knowledge-intensive task that requires significant time and effort to find related studies, identify evidence items, annotate the contents, and aggregate the results [1]. BrainKB, which stores knowledge from diverse data sources, including scientific publications, facilitates the meta-analysis.

	[1] _Tiddi, I., Balliet, D. and ten Teije, A., 2020. Fostering scientific meta-analyses with knowledge graphs: a case-study. In The Semantic Web: 17th International Conference, ESWC 2020, Heraklion, Crete, Greece, May 31–June 4, 2020, Proceedings 17 (pp. 287-303). Springer International Publishing._

## Models
- [GARS](https://github.com/brain-bican/models/blob/main/linkml-schema/gars.yaml)
- [ANSRS](https://github.com/brain-bican/models/blob/main/linkml-schema/ansrs_core.yaml)
- [Biolink](https://github.com/brain-bican/models/blob/main/linkml-schema/bican_biolink.yaml)

> - CAP/CAS

Reviewer 2: these should be updated with the correct pointers and expanded names. we are removing `RS` from the names.

Author: Will expand the names and point to the documentation - [https://brain-bican.github.io/models/ ](https://brain-bican.github.io/models/ )

## Usage Scenario
**Actor:** Alice (Neuroscientists)

**Task:** Alice wants to know if they can gain new insights from their newly collected neuroscience data.

> **Precondition:** The dataset is usable, i.e., is not corrupted and is related to the neuroscience domain.

Reviewer 2: what is this dataset?

Author: I have updated the description.

# BrainKB Design Document
<img src="logo.png" width="100px" style="float:right; margin-top:-45px"/>

**Author:** Tek Raj Chhetri | <tekraj@mit.edu>

## Overview

> _BrainKB serves as a knowledge base platform that provides scientists worldwide with tools for searching, exploring, and visualizing Neuroscience knowledge represented by knowledge graphs (KGs)_. Moreover, BrainKB provides cutting-edge tools that enable scientists to contribute new information (or knowledge) to the platform, ensuring it remains the go-to destination for all neuroscience-related research needs.

Reviewer: since this is brand new platform I would avoid saying "ensuring it remains the go-to destination"

Author: updated - "...is expected to be a go-to destination..."

> The main objective of BrainKB is to represent Neuroscience knowledge as a knowledge graph such that it can be used for different downstream tasks, such as making predictions and new inferences in addition to querying and viewing information.  The expected outcome of the BrainKB includes the following:

Reviewer: is there a reason why Neuroscience is always capitalized?

Author: no, fixed the issue

> - (Semi-)Automated extraction of Neuroscience knowledge from structured, semi-structured, and unstructured sources, representing the extracted knowledge via KGs.

Reviewer: ```suggestion
- (Semi-)Automated extraction of neuroscience knowledge from structured (e.g., relational database), semi-structured (e.g., JSON, XML and CSV files), and unstructured sources (e.g., text or video), and representing the knowledge via KGs.
```

- **Lack of Support for Heterogeneous Data Sources:** The current platforms in neuroscience are limited in their ability to handle a diverse range of data sources. For instance, [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008) can only integrate knowledge from 41 databases, whereas [BrainKnow](http://www.brain-knowledge-engine.org/) solely focuses on scientific literature. <span style="color: red;">However, knowledge is not restricted to just databases or scientific literature, and there is a need for platforms that can accommodate a wider variety of sources (e.g., structured, semi-structured and unstructured sources).</span>

> 

Reviewer 3: I think that above overview would be very incomplete without mentioning/comparing to efforts in EU/Ebrains:
- https://github.com/openMetadataInitiative for metadata harmonization
- their "knowledge graph" : https://www.ebrains.eu/data/find-data/find-data/

Author: Thanks for the pointer, I will update the document accordingly.

Author: Updated!

	- **Example:**
		- Structured data: Relational database
		- Semi-structured data: JSON, XML and CSV files
		- Unsuctured data: Text, Video files

## Principles

### Data Ingestion
BrainKB will support the data from various sources in different formats  (e.g., texts, JSON (JavaScript Object Notation)) for knowledge extraction via the BrainKB user interface (UI) and the API endpoints. Both batch and streaming data ingestion modes will be supported.
 

### Schema Flexibility

> KGs evolve over time. For example, if we consider the case of the president of a country, it changes overtime. The KGs storing the storing the information regarding the president of the country has to be updated accordingly. Similar is the case for the neuroscience or any other domain. The knowledge may change over time based on new research findings, thereby making previous knowledge obsolete or factually incorrect. Additionally, changes might also occur in the case of schema, such as due to the standardization or alignment or updates. While schema changes may not always be necessary, they may be required to accommodate new information. Therefore, BrainKB will support this evolution by allowing the addition (or removal) of entities and relationships (or new knowledge).

Reviewer 3: ```suggestion
KGs evolve over time. For example, if we consider the case of the president of a country, it changes overtime. The KGs storing the information regarding the president of the country has to be updated accordingly. Similar is the case for the neuroscience or any other domain. The knowledge may change over time based on new research findings, thereby making previous knowledge obsolete or factually incorrect. Additionally, changes might also occur in the case of schema, such as due to the standardization or alignment or updates. While schema changes may not always be necessary, they may be required to accommodate new information. Therefore, BrainKB will support this evolution by allowing the addition (or removal) of entities and relationships (or new knowledge).
```

FWIW might be best to adhere to "1 sentence per line" convention

- CAP/CAS

## Usage Scenario
**Actor:** Alice (Neuroscientists)

**Task:** Alice wants to know if they can gain new insights from their newly collected neuroscience data.

**Precondition:** The dataset is usable, i.e., is not corrupted and is related to the neuroscience domain.

**Flow:**

1. Alice uploads the data into the BrainKB platform through the BrainKB UI (User Interface).
2. BrainKB, the system, then analyzes data. If any error, e.g., unsupported file format, it will return the error; otherwise, the system will proceed to the next step of knowledge extraction.
3. The system will perform the knowledge extraction, validation, and alignment operation. If the validation or the alignment issue cannot be resolved automatically, the extracted knowledge represented via KG is flagged for expert review. Upon the successful review, the KGs are integrated (or stored) in the BrainKB storage and is available for visualization and analysis.

**Postcondition:** Alice discovers new insights through the integration of diverse knowledge sources represented in BrainKB's KGs.

## Technology
- FastAPI
- Docker

> - Serverles

Reviewer 3: ```suggestion
- Serverless
```

I will propose codespell ... ;-)

Author: Thanks, will fix it. Probably, I should stop using notepad.

# BrainKB Design Document
<img src="logo.png" width="100px" style="float:right; margin-top:-45px"/>

**Author:** Tek Raj Chhetri | <tekraj@mit.edu>

## Overview
_BrainKB serves as a knowledge base platform that provides scientists worldwide with tools for searching, exploring, and visualizing Neuroscience knowledge represented by knowledge graphs (KGs)_. Moreover, BrainKB provides cutting-edge tools that enable scientists to contribute new information (or knowledge) to the platform and is expected to be a go-to destination for all neuroscience-related research needs.

> The main objective of BrainKB is to represent neuroscience knowledge as a knowledge graph such that it can be used for different downstream tasks, such as making predictions and new inferences in addition to querying and viewing information.  The expected outcome of the BrainKB includes the following:

Reviewer: I've added couple suggestions to the bullet points. Some changes are only to keep the grammatical consistency. However I'm also not completely sure about the usage of term "extraction of KGs", but perhaps I'm just not familiar with the KG language, feel free to keep it

Author: I have merged your suggested text.

- (Semi-)Automated extraction of Neuroscience knowledge from structured, semi-structured, and unstructured sources, representing the extracted knowledge via KGs.
 
- Visualization of the KGs.

> - Provides features to perform different analytics operations over the BrainKB KGs.

Reviewer: ```suggestion
- Platform to perform different analytics operations over the BrainKB KGs.
```

> - (Semi-)automatically validates (e.g., quality) the extracted KGs to ensure that the BrainKB KGs are high quality. 

Reviewer: ```suggestion
- (Semi-)Automated validation of the BrainKGs to ensure the high quality of the content. 
```

> - Provides the ability to ingest data in batch or streaming mode for the automated extraction of KGs.

Reviewer: ```suggestion
- Tools to ingest data in batch or streaming mode for automated and efficient extraction of the information.
```

## Why BrainKB?

> - **Limited Availability of Platforms for Integrating Neuroscience Data into Knowledge Graphs:** In fields such as biomedicine, many platforms exist, such as, [SPOKE](https://doi.org/10.1093/bioinformatics/btad080) and [CIViC](https://civicdb.org/welcome). <span style="color: red;">However, such resources are comparatively limited in the domain of neuroscience.</span> [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008), a web-based platform that integrates anatomical, functional, and genetic knowledge, is among the limited number of such resources. [BrainKnow](http://www.brain-knowledge-engine.org/), the most recent platform, is another platform that is designed to [synthesizes and integrates neuroscience knowledge from scientific literature](https://arxiv.org/pdf/2403.04346.pdf). Additionally, projects like [DANDI](https://dandiarchive.org/), [EBRAINS](https://www.ebrains.eu/data/find-data/find-data/) and [Open Metadata Initiative](https://github.com/openMetadataInitiative) are making strides by enabling sharing of neurophysiology data together with its metadata. 

Reviewer: ```suggestion
- **Limited Availability of Platforms for Integrating Neuroscience Data into Knowledge Graphs:** In fields such as biomedicine, many platforms exist, such as, [SPOKE](https://doi.org/10.1093/bioinformatics/btad080) and [CIViC](https://civicdb.org/welcome). <span style="color: red;">However, such resources are comparatively limited in the domain of neuroscience.</span> [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008), a web-based platform that integrates anatomical, functional, and genetic knowledge, is among the limited number of such resources. [BrainKnow](http://www.brain-knowledge-engine.org/), the most recent platform, is another platform that is designed to [synthesize and integrate neuroscience knowledge from scientific literature](https://arxiv.org/pdf/2403.04346.pdf). Additionally, projects like [DANDI](https://dandiarchive.org/), [EBRAINS](https://www.ebrains.eu/data/find-data/find-data/) and [Open Metadata Initiative](https://github.com/openMetadataInitiative) are making strides by enabling sharing of neurophysiology data together with its metadata. 
```

- **Lack of Support for Heterogeneous Data Sources:** The current platforms in neuroscience are limited in their ability to handle a diverse range of data sources. For instance, [LinkRBrain](https://doi.org/10.1016/j.jneumeth.2014.12.008) can only integrate knowledge from 41 databases, whereas [BrainKnow](http://www.brain-knowledge-engine.org/) solely focuses on scientific literature. <span style="color: red;">However, knowledge is not restricted to just databases or scientific literature, and there is a need for platforms that can accommodate a wider variety of sources (e.g., structured, semi-structured and unstructured sources).</span>

	- **Example:**

> 		- Structured data: Relational database

Reviewer: I would remove this section and add it to the sentence in the overview (see suggested changes above)

Author: I have removed the example as suggested.

		- Semi-structured data: JSON, XML and CSV files
		- Unsuctured data: Text, Video files

## Principles

### Data Ingestion

> BrainKB will support the data from various sources in different formats  (e.g., texts, JSON (JavaScript Object Notation)) for knowledge extraction via the BrainKB user interface (UI) and the API endpoints. Both batch and streaming data ingestion modes will be supported.

Reviewer: ```suggestion
BrainKB will support knowledge extraction from various sources in different data formats  (e.g., texts, JSON (JavaScript Object Notation)) via the BrainKB user interface (UI) and the application programming interface (API) endpoints. Both batch and streaming data ingestion modes will be supported.
```

 

### Schema Flexibility

> KGs evolve over time. For example, if we consider the case of the president of a country, it changes overtime. The KGs storing the information regarding the president of the country has to be updated accordingly. Similar is the case for the neuroscience or any other domain. The knowledge may change over time based on new research findings, thereby making previous knowledge obsolete or factually incorrect. Additionally, changes might also occur in the case of schema, such as due to the standardization or alignment or updates. While schema changes may not always be necessary, they may be required to accommodate new information. Therefore, BrainKB will support this evolution by allowing the addition (or removal) of entities and relationships (or new knowledge).

Reviewer: ```suggestion
KGs evolve over time. For example, if we consider the case of the president of a country, it changes overtime. The KGs storing the information regarding the president of the country has to be updated accordingly. There are  many similar cases in neuroscience or any other domain. The knowledge may change over time based on new research findings, thereby making previous knowledge obsolete or factually incorrect. Additionally, changes might also occur in the schema due to the standardization, alignment or updates. While schema changes may not always be necessary, they may be required to accommodate new information. Therefore, BrainKB will support the evolution by allowing the addition (or removal) of entities and relationships.
```

> - **Example:** The concepts may change over time, for example, due to major guideline changes or adoption of different standards, or they might become obsolete. For instance, in fields like biology, [newer findings can invalidate existing terms](https://wiki.geneontology.org/Principles_for_term_obsoletion), requiring flexibility in the schema to account for future changes. 

Reviewer: I think in this "example" section there is too much repetition from the section above. Perhaps you can extend the section above with the example?

Author: I have updated the example to remove the repeated text.

> ### Maintainability

Reviewer: this section is too vague, what does it mean "validation to be performed easily"? perhaps (semi-)automatically? regularly? who will be responsible for maintainance?

Author: I have added the description.

BrainKB shall be maintainable, allowing operations such as KG enrichment and validation to be performed easily.

### Curation
BrainKB will allow the community-driven curation of the KGs as well as (semi-) automated extraction and construction of KGs from external sources, e.g, scientific literatures.

### Accuracy, Completeness and Consistency (ACC)

> BrainKB shall ensure the accuracy of the knowledge for which multi-step (semi-) automated validations will be performed. Additionally, checks will also be performed to ensure that the KG triples are complete, i.e., the mandatory information is present. Further to accuracy and completeness, BrainKB shall ensure that adding the new facts (or KG triples) will not lead to inconsistency (see figure below) with existing knowledge due to factual errors, data inconsistencies, and incompleteness.

Reviewer: ```suggestion
BrainKB shall check the accuracy of the knowledge for which multi-step (semi-) automated validations will be performed. Additionally, checks will be performed to ensure that the KG triples are complete, i.e., the mandatory information is present. Further to accuracy and completeness, BrainKB shall ensure that adding the new facts (or KG triples) will not lead to inconsistency (see figure below) with existing knowledge due to factual errors, data inconsistencies, and incompleteness.
```
I believe "ensuring" the accuracy of all the information is impossible, since the sources might not be correct

![](acc.png)

> _Figure 1: KGs. The image on the left shows the original knowledge graph, while the image on the right demonstrates the updated knowledge graph. The green highlighted box indicates new knowledge that has been added, while the **red highlighted box** indicates any inconsistencies caused by factual changes, i.e., **incorrect date of birth**._

Reviewer: I'm sorry but IMO this figure is not clear, since it suggested that only one triple is added (the green one), so why suddenly  the birth is incorrect?

Author: I have updated the figure caption based on Yarik suggestion to make explicit why the one highlighted with red is incorrect.

> The ACC process will ensure human-centricity is maintained alongside automated validation.

Reviewer: I would either remove this sentence or tell something more what you mean by this...

Author: Updated the description.

### Provenance

> To enable trust, the provenance, such as the source of the information and the curators (in the case of manual) of all the information, shall be maintained. The provenance conflict resolution mechanism will also be implemented to ensure the accuracy of the provenance information.

Reviewer: ```suggestion
To enable trust, the provenance, i.e., documentation of the source and the curators (in case of manual curation) of all the information, shall be maintained. The provenance conflict resolution mechanism will also be implemented to ensure the accuracy of the provenance information.
```

### Querying and Reasoning
BrainKB shall support the KGs' querying and reasoning. It shall also support other downstream analytics tasks, such as link predictions using machine learning techniques.

### Integration and Interoperability

> To ensure interoperability and ease of integration, BrainKB will focus on using standardized ontologies or schemas. However, not all standardized ontologies or schemas are available. In such cases, other schemas or ontologies must be used. To ensure the interoperability, the alignment will be performed where necessary.

Reviewer: ```suggestion
To ensure interoperability and ease of integration, BrainKB will focus on using standardized ontologies or schemas. However,  standardized ontologies or schemas are not always available. In such cases, other schemas or ontologies must be used. To ensure the interoperability, the alignment will be performed where necessary.
```

### Minimize Cognitive Burden and Data Fatigue

> As BrainKB will also provide features to perform the analytics operation in addition to querying the information (or knowledge), a special emphasis shall be placed on ensuring that the information presented to the user does not cause a cognitive burden and data fatigue. A cognitive burden occurs when the brain must exert more effort to understand information, typically resulting from an overload of visual content. For example, the figure below (left) places more cognitive burden than on the right.

Reviewer: I have to say that this is not obvious to me, since the "cognitive burden" depends not only on the amount of the information on the figure but also on the audience background knowledge, but perhaps this is discussion for other time.
The figure is completely unreadable, both panels

Author: Yes, but I was talking from the presentation perspective.

![](cognitive_burden.png)

## Features

## Other considerations

> __Assumption:__ We operate on open-world assumptions (OWA), not closed-world assumptions (CSA). In OWA, we do not make any assumptions about the absence of statements, while in CSA absence of statements would be evaluated as false, i.e., assumed to be false.

Reviewer: perhaps you can provide some example?

Author: added example

## Architecture
The figure below shows the high-level overview of the components of the BrainKB architecture.

![](ingest.png)

> _Figure 2: Ingest Service_

Reviewer: The figure 2 has no explanation, and it is not clear how it relates to figure 3

Author: added description

> __Application:__ The application (or the application layer) is the go-to point that provides access to BrainKB, such as via UI.

Reviewer: There is no "Application" section in the fig.3

Author: Updated the description.

__Service:__ The service layer implements the core logic and is broken down into multiple services based on the functionalities.  
__Resource:__ The resource will provide the necessary computational resources that are required to deliver the required service by BrainKB.

![](initial-arch.png)
_Figure 3: Architecture of BrainKB_

> ## Target Audience

Reviewer: perhaps you should use future tense in this section

Author: updated description

- **Neuroscience researchers:** BrainKB's primary audience is neuroscience researchers, who can use the platform to integrate, visualize, and analyze neuroscience data. They can also capitalize on the platform's ability to synthesize their data (or knowledge) into KGs.

> - **Research Labs and Academic Institutions:** BrainKB will be an invaluable resource for teaching and research in academic contexts specializing in neuroscience research. It offers convenient access to integrated neuroscience data for faculty and students.

Reviewer: since the previous point is about research, perhaps this point can focus on the education/teaching only and do not repeat the research?

Author: updated

- **Policy Makers and Healthcare Professionals:** Neurology policymakers can use BrainKB, aka the neuroscience knowledge that BrainKB hosts, to make policy decisions. Furthermore, healthcare professionals in neurology (or clinical neuroscience) may also use BrainKB knowledge to understand and improve neurological disease outcomes.

- **Neuroscience-related Companies:** Companies specializing in developing drugs for neurological diseases can use the platform's KGs to gain insights into neurological conditions and treatments.

## Use cases

> - **Extraction/Integration/Refinement:** BrainKB will provide features to extract knowledge from diverse sources, such as raw text and scientific publications, and integrate it with the knowledge represented via KGs. Additionally, BrainKB also provides features to refine the extracted knowledge, e.g., through humans in the loop.

Reviewer: ```suggestion
- **Extraction/Integration/Refinement:** BrainKB will provide features to extract knowledge from diverse sources, such as raw text and scientific publications, and integrate it with the knowledge represented via KGs. Additionally, BrainKB will also provides features to refine the extracted knowledge, e.g., through humans in the loop.
```

> - **Cards:** The BrainKB web application allows easy visualization of the knowledge of interest to scientists/researchers stored in KGs and their corresponding interconnected knowledge. The figure below (Figure 4) shows a snippet of the entity card from the BrainKB web application, which can be accessed at [http://beta.brainkb.org](http://beta.brainkb.org).

Reviewer: ```suggestion
- **Cards:** The BrainKB web application allows easy visualization of the knowledge of interest to scientists/researchers stored in KGs and their corresponding interconnected knowledge. Figure 4 shows a snippet of the entity card from the BrainKB web application, which can be accessed at [http://beta.brainkb.org](http://beta.brainkb.org).
```
The first sentence is pretty complex, but not sure how to change it. Is this figure really about "knowledge" or just data/information?

- **Find/correct Errors:** BrainKB will provide a feature to search existing knowledge and correct errors if any. 

- **Add information/API:** BrainKB offers an API endpoint that enables seamless integration with its platform. These endpoints facilitate data ingestion from various sources, such as CSV files or raw text, for constructing KGs, performing search operations, and conducting analyses on the stored KGs.

- **Doing meta-analysis:** Meta-analysis is a knowledge-intensive task that requires significant time and effort to find related studies, identify evidence items, annotate the contents, and aggregate the results [1]. BrainKB, which stores knowledge from diverse data sources, including scientific publications, facilitates the meta-analysis.

	[1] _Tiddi, I., Balliet, D. and ten Teije, A., 2020. Fostering scientific meta-analyses with knowledge graphs: a case-study. In The Semantic Web: 17th International Conference, ESWC 2020, Heraklion, Crete, Greece, May 31–June 4, 2020, Proceedings 17 (pp. 287-303). Springer International Publishing._

## Models

> - Genome Annotation Registry Service (GARS) Model

Reviewer: ```suggestion
Models currently used in BrianKB:

- Genome Annotation Registry Service (GARS) Model
```

- Anatomical Structure Reference Service (AnSRS) Model
- Library Generation Model

Detailed descriptions of the models above are available at [https://brain-bican.github.io/models/](https://brain-bican.github.io/models/). 

> ## Usage Scenario

Reviewer: perhaps this should be right after the "Audience" section?

Author: yes

# BrainKB Design Document
<img src="logo.png" width="100px" style="float:right; margin-top:-45px"/>

**Author:** Tek Raj Chhetri | <tekraj@mit.edu>

## Overview
_BrainKB serves as a knowledge base platform that provides scientists worldwide with tools for searching, exploring, and visualizing Neuroscience knowledge represented by knowledge graphs (KGs)_. Moreover, BrainKB provides cutting-edge tools that enable scientists to contribute new information (or knowledge) to the platform and is expected to be a go-to destination for all neuroscience-related research needs.

The main objective of BrainKB is to represent neuroscience knowledge as a knowledge graph such that it can be used for different downstream tasks, such as making predictions and new inferences in addition to querying and viewing information.  The expected outcome of the BrainKB includes the following:

- (Semi-)Automated extraction of neuroscience knowledge from structured, semi-structured, and unstructured sources, and representing the knowledge via KGs.
 
- Visualization of the KGs.

- Platform to perform different analytics operations over the BrainKB KGs.

> - (Semi-)Automated validation of the BrainKGs to ensure the high quality of the content. 

Reviewer 4: ```suggestion
- (Semi-)Automated validation of the BrainKB KGs to ensure the high quality of the content. 
```

To be consistent with the wording in the preceding point.

- **Example:** In fields like biology, [newer findings can invalidate existing terms](https://wiki.geneontology.org/Principles_for_term_obsoletion), requiring flexibility in the schema to account for future changes. 

### Maintainability
BrainKB shall be maintainable, allowing operations such as KG enrichment and validation to be performed easily. When we mention "validation to be performed easily," we are referring to processes that require minimal human intervention, either through semi-automated or fully-automated methods.
### Curation
BrainKB will allow the community-driven curation of the KGs as well as (semi-) automated extraction and construction of KGs from external sources, e.g, scientific literatures.

### Accuracy, Completeness and Consistency (ACC)

> BrainKB shall check the accuracy of the knowledge for which multi-step (semi-) automated validations will be performed. Additionally, checks will be performed to ensure that the KG triples are complete, i.e., the mandatory information is present. Further to accuracy and completeness, BrainKB shall ensure that adding the new facts (or KG triples) will not lead to inconsistency (see figure below) with existing knowledge due to factual errors, data inconsistencies, and incompleteness.

Reviewer 4: ```suggestion
BrainKB shall check the accuracy of the knowledge for which multi-step (semi-) automated validations will be performed. Additionally, checks will be performed to ensure that the KG triples are complete, i.e., the mandatory information is present. To further ensure accuracy and completeness, BrainKB shall guarantee the additions of new facts (or KG triples) will not lead introduce inconsistencies (see figure below) with existing knowledge due to factual errors, data inconsistencies, and incompleteness.
```

**Example:** Let's consider a university scenario. We want to determine if Jane Doe is enrolled in the **AI 101** course.
	
- In CSA, if Jane Doe's enrollment information for AI 101 is not present in the university database, this absence of information is interpreted as Jane Doe not being enrolled in the course.
- Conversely under OWA, the absence of Jane Doe's enrollment information for AI 101 means that the information is simply missing and it remains uncertain whether Jane Doe is enrolled or not enrolled in the course.

## Architecture
The figure below (Figure 3) shows the BrainKB's architecture. It is divided into three layers: application (layer 1), service (layer 2), and resource (layer 3).

__Application:__ The application (or the application layer or layer 1) is the go-to point that provides access to BrainKB, such as via UI.

__Service:__ The service layer (or layer 2) implements the core logic and is broken down into multiple services based on the functionalities.  

> __Resource:__ The resource (or layer 3) will provide the necessary computational resources that are required to deliver the required service by BrainKB.

Reviewer 4: ```suggestion
__Resource:__ The resource (or layer 3) will provide the necessary computational resources that are required to deliver the required services by BrainKB.
```

![](brainkb-arch.png)
_Figure 3: Architecture of BrainKB_

Additionally, Figure 4, shows the architecture of the ingest service, a component of BrainKB.

> ![](ingest.png)

Reviewer 4: This figure is a bit confusing to me. It seems to show that BrainKB is a component of the ingest service instead of the other way around.

Author: yes, I will remove the BrainKB component from the Ingest service

Author: I have updated the image.

_Figure 4: Ingest Service_

## Target Audience

- **Neuroscience researchers:** BrainKB's primary audience is neuroscience researchers, who can use the platform to integrate, visualize, and analyze neuroscience data. They can also capitalize on the platform's ability to synthesize their data (or knowledge) into KGs.

- **Research Labs and Academic Institutions:** BrainKB will be an invaluable resource for teaching and research in academic contexts specializing in neuroscience research. It offers convenient access to integrated neuroscience data for faculty and students.

> - **Policy Makers and Healthcare Professionals:** Neurology policymakers can use BrainKB, aka the neuroscience knowledge that BrainKB hosts, to make policy decisions. Furthermore, healthcare professionals in neurology (or clinical neuroscience) may also use BrainKB knowledge to understand and improve neurological disease outcomes.

Reviewer 4: ```suggestion
- **Policy Makers:** Neurology policymakers can use the neuroscience knowledge that BrainKB hosts to make policy decisions. 
- **Healthcare Professionals:** Healthcare professionals in neurology (or clinical neuroscience) may use BrainKB knowledge to understand and improve neurological disease outcomes.
```

- **Neuroscience-related Companies:** Companies specializing in developing drugs for neurological diseases can use the platform's KGs to gain insights into neurological conditions and treatments.

## Usage Scenario
**Actor:** Alice (Neuroscientists)

> **Task:** Alice wants to know if he can gain new insights from their newly collected neuroscience data.

Reviewer 4: ```suggestion
**Task:** Alice wants to know if she can gain new insights from their newly collected neuroscience data.
```

**Precondition:** The newly collected neuroscience dataset, which includes demographics, gene expression maps, and structural and functional MRI scans, is usable and uncorrupted.

**Flow:**

1. Alice uploads the data into the BrainKB platform through the BrainKB UI (User Interface).
2. BrainKB, the system, then analyzes data. If any error, e.g., unsupported file format, it will return the error; otherwise, the system will proceed to the next step of knowledge extraction.
3. The system will perform the knowledge extraction, validation, and alignment operation. If the validation or the alignment issue cannot be resolved automatically, the extracted knowledge represented via KG is flagged for expert review. Upon the successful review, the KGs are integrated (or stored) in the BrainKB storage and is available for visualization and analysis.

**Postcondition:** Alice discovers new insights through the integration of diverse knowledge sources represented in BrainKB's KGs.

## Use Cases

> - **Extraction/Integration/Refinement:** BrainKB will provide features to extract knowledge from diverse sources, such as raw text and scientific publications, and integrate it with the knowledge represented via KGs. Additionally, BrainKB will also provides features to refine the extracted knowledge, e.g., through humans in the loop.

Reviewer 4: ```suggestion
- **Extraction/Integration/Refinement:** BrainKB will provide features to extract knowledge from diverse sources, such as raw text and scientific publications, and integrate it with the knowledge represented via KGs. Additionally, BrainKB will also provide features to refine the extracted knowledge, e.g., through humans in the loop.
```

            Alignment-->>+ Validation: Return response (triples with resolved conflicts)
            

        else conflict identified, perform resolution-requires human oversight
            Alignment->>+ Expert: Send to expert for manual conflict resolution
            Expert-->>+Alignment: Return response (triples with resolved conflicts)
            Alignment-->>+ Validation: Return response (triples with resolved conflicts) 
            
        end 
            Validation-->>+ Mapping: Return response (validated and conflict resolved KG triples)
            Mapping-->>+ KG_Construction: Updated KG triples
            KG_Construction->>+Triplestore: Store KG in database
            Triplestore-->>+KG_Construction: Return acknowledgement
            KG_Construction-->>+UI: Return response (operation status notification)
            UI-->>+User:Send notification
    end
```

## Hosting Infrastructure

We recognize the current [beta site’s](https://beta.brainkb.org/) issues and are working towards improving it. In particular, we are working on improving the following problems of the beta site.

> - **Performance:** There is a performance issue with the beta site's knowledge base page. This is because we are currently using the free version of GraphDB and it only allows two simultaneous queries. Since, fetching the details, such as Library Aliquot and their inter-related information requires running more than two queries, there are performance issues.  We are looking for the premium version of GraphDB and other opensource triple stores as alternatives.

Reviewer 4: ```suggestion
- **Performance:** We are currently using the free version of GraphDB. It has a limitation of two simultaneous queries. Because fetching the details, such as Library Aliquot and their inter-related information, requires running more than two queries, this limitation impacts the beta site's knowledge base page.  We are currently considering the premium version of GraphDB and other opensource triple stores alternatives.
```

Possibly, you may want to added something like ", and the performance of future version of BrainKB will be significantly improved."

Author: Updated

## Github Repository 

- **Source code**
	- [https://github.com/sensein/BrainKB](https://github.com/sensein/BrainKB) - Backend services
	- [https://github.com/sensein/brainkb-ui/tree/admin-ui](https://github.com/sensein/brainkb-ui/tree/admin-ui) - UI
- **Developer documentation**
	- [https://github.com/sensein/brainkbdocs](https://github.com/sensein/brainkbdocs) 	
	
## Work Plan 

| Features       | Status                          |
|------------|--------------------------------|
| [UI with NextJS](https://github.com/sensein/brainkb-ui) | Implementation in progress       |
| KG construction from scientific publication  | Designed the approach and implementation in progress |
| Structured models (or ontology) design | Implementation in progress|

> | [BrainKB documentation including deployment instructions and lessons learned](https://sensein.group/brainkbdocs/) | Pratially complete. Updates will be made as the work progress   |

Reviewer 4: ```suggestion
| [BrainKB documentation including deployment instructions and lessons learned](https://sensein.group/brainkbdocs/) | Practically complete. Updates will be made as the work progresses   |
```
